R/string_boundary_linter.R

```diff
@@ -0,0 +1,87 @@
+#' Require usage of startsWith() and endsWith() over grepl()/substr() versions
+#'
+#' [startsWith()] is used to detect fixed initial substrings; it is more
+#'   readable and more efficient than equivalents using [grepl()] or [substr()].
+#'   c.f. `startsWith(x, "abc")`, `grepl("^abc", x)`,
+#'   `substr(x, 1L, 3L) == "abc"`.
+#'
+#' Ditto for using [endsWith()] to detect fixed terminal substrings.
+#'
+#' @export
+string_boundary_linter <- function() {
+  anchor_expr <- xp_or(
+    "substring(text(), 2, 1) = '^'",
+    "substring(text(), string-length(text()) - 1, 1) = '$'"
+  )
+  ignore_case_expr <- xp_and(
+    "text() = 'ignore.case'",
+    "not(following-sibling::expr[1][NUM_CONST[text() = 'FALSE']])"
+  )
+  grepl_expr <- xp_and(
+    "preceding-sibling::expr[1][SYMBOL_FUNCTION_CALL[text() = 'grepl']]",
+    sprintf("not(parent::expr[SYMBOL_SUB[%s]])", ignore_case_expr)
+  )
+  str_detect_expr <- file.path(
+    "parent::expr",
+    "preceding-sibling::expr[2][SYMBOL_FUNCTION_CALL[text() = 'str_detect']]"
+  )
+  str_const_cond <- xp_and(
+    anchor_expr,
+    "string-length(text()) > 3",
+    xp_or(sprintf("parent::expr[%s]", grepl_expr), str_detect_expr)
+  )
+  regex_xpath <- sprintf("//STR_CONST[%s]", str_const_cond)
```

Comment:
This is very convoluted.
Can you simplify the generation and may be `glue()` the conditions for easier readability?

Also note the substring XPaths preclude raw strings.

Reply:
good call -- way simpler now & handles raw strings